### ISSUE TITLE
Analyzer to suggest async if Task.Wait()/Result/GetAwaiter().GetResult() is used

### ISSUE THREAD

**Issue by jamesqo** [xunit/xunit#1390]
This was suggested by [a user on Reddit](https://www.reddit.com/r/csharp/comments/6rigmd/what_are_some_common_mistakes_you_make_with/dl5awsk/). Since xUnit supports async methods, there's no reason to not use `await` instead.

Also related: [this user](https://www.reddit.com/r/csharp/comments/6rigmd/what_are_some_common_mistakes_you_make_with/dl5dkx9/) suggests we should issue a warning if `async void` is used instead of `async Task`, to improve test performance.

**Comment by marcind** [comment]
I've thought about having something detect patterns that wait on async. I've been on the fence about it.

Reason against:
- I think there's probably a need for a generic analyzer that points out those usages (and in fact, at a previous job we had exactly that). I imagine it's already been implemented in the public domain and so not sure if it's worth replicating that work.

Reason for:
- XUnit users might not have this generic analyzer and/or they might not be aware that xunit supports async test methods

Probably some investigation is warranted about whether there's something like that out there already. Since using async/await as opposed to spin-waiting in tests should result in faster test execution I'm in general not opposed to adding this into the analyzers, assuming it won't duel with whatever else is in the community.

**Comment by marcind** [comment]
Moving discussion of `async void` to https://github.com/xunit/xunit/issues/1405

**Comment by jamesqo** [comment]
@marcind 

> I think there's probably a need for a generic analyzer that points out those usages (and in fact, at a previous job we had exactly that). I imagine it's already been implemented in the public domain and so not sure if it's worth replicating that work.

There are indeed some analyzers already out there related to `async`. Related to this issue, I found [Async006](https://github.com/dotnet/roslyn-analyzers/blob/569bd373a4831d3035597197e02980b57602a7f2/src/Unfactored/AsyncPackage/AsyncPackage.md#async006-dont-mix-blocking-and-async) which warns against using blocking calls inside a method that is already async.

However, I don't think there's an analyzer out there that will tell you to *convert* a synchronous method to an async one if it detects a blocking call inside the method. Obviously, that would break a lot of things because all callsites of the method would need to become `async` and use `await`, etc. In the context of xUnit.net, however, it's fine to convert a sync test method to an async one because the runner can handle both.

I suggest either of the options below:

```cs
// Case 1
[Fact]
public void TestSync()
{
    int result = Task.FromResult(3).Result;
}

// Case 2
[Fact]
public async void TestAsync()
{
    int result = Task.FromResult(3).Result;
}
```

**Option 1.** Warn/offer codefix for Case 1, but not Case 2. Pros: We wouldn't be duplicating the work of Async006.
**Option 2.** Warn/offer codefix for both Case 1 and Case 2. Pros: Async006 [isn't even implemented](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/DoNotMixBlockingAndAsync.cs) yet and xUnit users might not have the generic analyzers.

**Comment by wachulski** [comment]
@jamesqo @marcind 

In xunit/xunit.analyzers#119 I follow `Option 2`, but in broader sense. The warning is shown wherever in the assembly with xUnit (for each call of spin-waiting Task members).

Rationale:
1. I see 2 groups of xUnit users: advanced (`A`) and beginners (`B`). `A`s know details of async/Task/SyncContext/etc. and what potential pitfalls are. `B`s don't.
2. I assume `size of A << size of B`.
3. Each advanced user can opt-out the rule and replace it with a dedicated async analyzer.
4. Conclusion: this little analysis aid would help much more than cause irritation.

**Comment by bradwilson** [comment]
I've decided to just show an error when using a problematic method or property. That includes

* `Task.GetAwaiter().GetResult()`
* `Task.Wait()`
* `Task.WaitAll()`
* `Task.WaitAny()`
* `Task<T>.GetAwaiter().GetResult()`
* `Task<T>.Result`
* `ValueTask.GetAwaiter().GetResult()`
* `ValueTask<T>.GetAwaiter.GetResult()`
* `ValueTask<T>.Result`

I'm not going to add a fixer here, because of the complexity involved. The message will tell the user what to do (use an async test method and `await`).

**Comment by bradwilson** [comment]
Shipped in `1.3.0-pre.12`
